From bbe6c655d209a6d46d856076f5d4d758f763d491 Mon Sep 17 00:00:00 2001
From: Ric Li <ming3.li@intel.com>
Date: Mon, 14 Feb 2022 16:28:01 +0800
Subject: [PATCH 1/7] pcapng: add ns timestamp for copy api

The original API only receives cycles and converts it to ns.
Add ns timestamp to directly write it to pcapng file.

Signed-off-by: Ric Li <ming3.li@intel.com>

Fix build on v22.10

Signed-off-by: Du, Frank <frank.du@intel.com>
---
 app/test/test_pcapng.c  | 4 ++--
 lib/pcapng/rte_pcapng.c | 5 +++--
 lib/pcapng/rte_pcapng.h | 2 +-
 lib/pdump/rte_pdump.c   | 2 +-
 4 files changed, 7 insertions(+), 6 deletions(-)

diff --git a/app/test/test_pcapng.c b/app/test/test_pcapng.c
index a7acbdc058..8e672a41f2 100644
--- a/app/test/test_pcapng.c
+++ b/app/test/test_pcapng.c
@@ -139,7 +139,7 @@ test_write_packets(void)
 		struct rte_mbuf *mc;
 
 		mc = rte_pcapng_copy(port_id, 0, orig, mp, pkt_len,
-				rte_get_tsc_cycles(), 0);
+				rte_get_tsc_cycles(), 0, 0);
 		if (mc == NULL) {
 			fprintf(stderr, "Cannot copy packet\n");
 			return -1;
@@ -255,7 +255,7 @@ test_write_over_limit_iov_max(void)
 		struct rte_mbuf *mc;
 
 		mc = rte_pcapng_copy(port_id, 0, orig, mp, pkt_len,
-				rte_get_tsc_cycles(), 0);
+				rte_get_tsc_cycles(), 0, 0);
 		if (mc == NULL) {
 			fprintf(stderr, "Cannot copy packet\n");
 			return -1;
diff --git a/lib/pcapng/rte_pcapng.c b/lib/pcapng/rte_pcapng.c
index 80d08e1a3b..02bd563c82 100644
--- a/lib/pcapng/rte_pcapng.c
+++ b/lib/pcapng/rte_pcapng.c
@@ -449,7 +449,7 @@ struct rte_mbuf *
 rte_pcapng_copy(uint16_t port_id, uint32_t queue,
 		const struct rte_mbuf *md,
 		struct rte_mempool *mp,
-		uint32_t length, uint64_t cycles,
+		uint32_t length, uint64_t cycles, uint64_t tm_ns,
 		enum rte_pcapng_direction direction)
 {
 	struct pcapng_enhance_packet_block *epb;
@@ -463,7 +463,8 @@ rte_pcapng_copy(uint16_t port_id, uint32_t queue,
 #ifdef RTE_LIBRTE_ETHDEV_DEBUG
 	RTE_ETH_VALID_PORTID_OR_ERR_RET(port_id, NULL);
 #endif
-	ns = pcapng_tsc_to_ns(cycles);
+	ns = tm_ns;
+	if(ns == 0)	ns = pcapng_tsc_to_ns(cycles);
 
 	orig_len = rte_pktmbuf_pkt_len(md);
 
diff --git a/lib/pcapng/rte_pcapng.h b/lib/pcapng/rte_pcapng.h
index 7d2697c647..c0fc0ccd81 100644
--- a/lib/pcapng/rte_pcapng.h
+++ b/lib/pcapng/rte_pcapng.h
@@ -110,7 +110,7 @@ __rte_experimental
 struct rte_mbuf *
 rte_pcapng_copy(uint16_t port_id, uint32_t queue,
 		const struct rte_mbuf *m, struct rte_mempool *mp,
-		uint32_t length, uint64_t timestamp,
+		uint32_t length, uint64_t timestamp, uint64_t tm_ns,
 		enum rte_pcapng_direction direction);
 
 
diff --git a/lib/pdump/rte_pdump.c b/lib/pdump/rte_pdump.c
index a81544cb57..9737a79ba3 100644
--- a/lib/pdump/rte_pdump.c
+++ b/lib/pdump/rte_pdump.c
@@ -122,7 +122,7 @@ pdump_copy(uint16_t port_id, uint16_t queue,
 		if (cbs->ver == V2)
 			p = rte_pcapng_copy(port_id, queue,
 					    pkts[i], mp, cbs->snaplen,
-					    ts, direction);
+					    ts, 0, direction);
 		else
 			p = rte_pktmbuf_copy(pkts[i], mp, 0, cbs->snaplen);
 
-- 
2.34.1

